### PR TITLE
Add shell completion

### PR DESCRIPTION
**Pull request by creepinson**
## Changes

I added shell completion for zsh, bash or any other shell that clap_generate supports.

This is how I used it in my setup with zsh:

```bash
youki completion -g zsh > ~/.zfunc/youki.zsh
```

### CODE REVIEW

**Comment by utam0k** [comment]
@creepinson 
Hi! Thanks for your suggestion. But you may  have forgotten running format.

**Comment by creepinson** [comment]
You're right, I had used vscode's rust-analyzer formatter assuming that would work fine.

File: crates/youki/src/commands/completion.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+use clap::{App, Parser};
+use clap_generate::{generate, Generator, Shell};
+use std::io;
+
+#[derive(Debug, Parser)]
```
Comment (Furisto) [inline crates/youki/src/commands/completion.rs:5]: youki -h currently does not have a description for this command

```suggestion
Generate scripts for shell completion
#[derive(Debug, Parser)]
```

File: crates/youki/src/commands/completion.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+pub struct CompletionParser {
+    #[clap(long = "generator", short = 'g', arg_enum)]
+    pub generator: Shell,
+}
+
+pub fn print_completions<G: Generator>(gen: G, app: &mut App) -> Result<(), anyhow::Error> {
```
Comment (utam0k) [inline crates/youki/src/commands/completion.rs:11]: How about using `anyhow::Result`?

File: crates/youki/src/main.rs [hunk 3]
```diff
@@ -100,6 +102,9 @@ fn main() -> Result<()> {
         },
 
         SubCommand::Info(info) => commands::info::info(info),
+        SubCommand::Completion(parser) => {
+            commands::completion::print_completions(parser.generator, &mut app)
+        }
```
Comment (Furisto) [inline crates/youki/src/main.rs:107]:
```suggestion
        SubCommand::Completion(completion) => {
            commands::completion::completion(completion, &mut app)
        }
```

File: crates/youki/src/commands/completion.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+use clap::{App, Parser};
+use clap_generate::{generate, Generator, Shell};
+use std::io;
+
+#[derive(Debug, Parser)]
+pub struct CompletionParser {
+    #[clap(long = "generator", short = 'g', arg_enum)]
+    pub generator: Shell,
+}
```
Comment (Furisto) [inline crates/youki/src/commands/completion.rs:9]:
```suggestion
pub struct Completion {
    #[clap(long = "shell", short = 's', arg_enum)]
    pub shell: Shell,
}
```

File: crates/youki/src/main.rs [hunk 5]
```diff
@@ -46,6 +46,7 @@ enum SubCommand {
 
     // Youki specific extensions
     Info(info::Info),
+    Completion(commands::completion::CompletionParser),
```
Comment (Furisto) [inline crates/youki/src/main.rs:49]:
```suggestion
    Completion(completion::Completion),
```

File: crates/youki/src/commands/completion.rs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+pub fn print_completions<G: Generator>(gen: G, app: &mut App) -> Result<(), anyhow::Error> {
+    generate(gen, app, app.get_name().to_string(), &mut io::stdout());
+
+    Ok(())
+}
```
Comment (Furisto) [inline crates/youki/src/commands/completion.rs:15]:
```suggestion
pub fn print_completions(args: Completion, app: &mut App) -> Result<()> {
    generate(args.shell, app, app.get_name().to_string(), &mut io::stdout());

    Ok(())
}
```

**Review by Furisto — CHANGES_REQUESTED** [review]

**Comment by Furisto** [comment]
Minor stuff, the style should be consistent with the other commands.

**Comment by creepinson** [comment]
Looks like I messed up the order of the comment and the `#[derive(Debug, Parser)]` line next to the struct, although `cargo fmt --all --check` is successful. Otherwise I believe I made the requested changes.

**Review by utam0k — APPROVED** [review]
Thanks!

**Review by Furisto — APPROVED** [review]

**Comment by Furisto** [comment]
Great!

**Pull request merged**